### PR TITLE
proxy protocol for sandwich

### PR DESCRIPTION
**Pull request by stevenctl**
We add `ApplicationTunnel` as an optional piece of `GatewayAddress` to allow more advanced sandwiches:

* It's possible that a sandwiched waypoint wants additional info like PROXY protocol
* It's possible that the sandwiched waypoint binds to a different port than the HBONE address targets

### CODE REVIEW

File: proto/workload.proto [hunk 1]
```diff
@@ -212,6 +212,30 @@ enum TunnelProtocol {
...
+message ApplicationTunnel {
+  enum Protocol {
+    // Bytes are copied from the inner stream without modification.
+    NONE = 0;
+
+    // Prepend PROXY protocol headers before copying bytes
+    // Standard PROXY source and destination information
+    // is included, along with potential extra TLV headers:
+    // 0xD0 - The SPIFFE identity of the source workload
+    // 0xD1 - The FQDN or Hostname of the targeted Service
```
Comment (ilrudie) [inline proto/workload.proto:227]: Is 0xD1 no longer used, automatically implemented by the crate or a TODO?

Comment (stevenctl) [reply proto/workload.proto:227]: This is planned in the future to allow routing by host instead of VIP.

Comment (ilrudie) [reply proto/workload.proto:227]: sounds good. TY

File: src/proxy/inbound.rs [hunk 2]
```diff
@@ -432,26 +469,43 @@ impl Inbound {
...
-            return Ok((next_hop, us_wl, us_svc));
-        }
+            (next_hop, us_wl, us_svc)
+        } else {
+            return Err(Error::IPMismatch(conn.dst.ip(), hbone_addr.ip()));
+        };
 
-        Err(Error::IPMismatch(conn.dst.ip(), hbone_addr.ip()))
+        // Application tunnel may override the port.
+        let (upstream_addr, inbound_protocol) = match upstream.application_tunnel.clone() {
```
Comment (howardjohn) [inline src/proxy/inbound.rs:496]: nit: just make Protocol `Copy` the you won't need the `clone()`

File: src/proxy/inbound.rs [hunk 3]
```diff
@@ -551,6 +605,12 @@ pub(super) enum InboundConnect {
     DirectPath(TcpStream),
     /// Hbone is a standard HBONE request coming from the network.
     Hbone(Request<Incoming>),
+    // PROXY adds source and dest headers and source identity before forwarding bytes.
+    Proxy(
+        Request<Incoming>,
+        (SocketAddr, SocketAddr),
+        Option<Identity>,
```
Comment (howardjohn) [inline src/proxy/inbound.rs:612]: nit: any reason to not just  pass the full `rbac::Connection`?

**Review by howardjohn — APPROVED** [review]

**Pull request merged**